# 1. Perkenalan

<b>Nama: Rohmad </b>

<b>Tanggal: </b>05/07/2021

<b>Materi: </b>Live Code

<b>Dataset:</b>Dataset dapat diakses pada link berikut: https://github.com/fahmimnalfrzki/Dataset/blob/main/NMC1-9.xlsx?raw=true

time(s): Detik ke- pengukuran
V_m: Voltage/tegangan yang terukur tiap waktu
Current (A): Kuat arus terukur tiap waktu
Cycle: Nomor siklus charge-discharge baterai
Cap: Kapasitas baterai tiap waktu

<b>Objective:</b>menghitung hambatan/resistansi dan Daya setiap detik

In [1]:
#import library
import numpy as np
import pandas as pd
print(np.__version__)
print(pd.__version__)

1.20.1
1.2.4


# Data Loading

Load data dan melihat data teratas

In [2]:
df = pd.read_excel("NMC1-9.xlsx")
df.head() #melihat data teratas

time(s)    V_m  Current (A)  Cycle       Cap
0        0  4.098        0.997      1  2.840817
1        1  4.096        0.998      1  2.840817
2        2  4.094        0.999      1  2.840817
3        3  4.093        0.998      1  2.840817
4        4  4.092        0.999      1  2.840817

melihat 5 data terbawah

In [59]:
df.tail() #melihat data terbawah

time(s)    V_m  Current (A)  Cycle      Cap
90977    90977  2.506        1.002      9  2.77018
90978    90978  2.504        1.002      9  2.77018
90979    90979  2.502        1.002      9  2.77018
90980    90980  2.500        1.002      9  2.77018
90981    90981  2.497        1.002      9  2.77018

melihat tipe data yang ada di dataset

In [7]:
df.info() #melihat jenis tipe data

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 90982 entries, 0 to 90981
Data columns (total 5 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   time(s)      90982 non-null  int64  
 1   V_m          90982 non-null  float64
 2   Current (A)  90982 non-null  float64
 3   Cycle        90982 non-null  int64  
 4   Cap          90982 non-null  float64
dtypes: float64(3), int64(2)
memory usage: 3.5 MB


mnegecek presentase yang ada didataset

In [12]:
df.isna().mean() #melihat presentase null

time(s)        0.0
V_m            0.0
Current (A)    0.0
Cycle          0.0
Cap            0.0
dtype: float64

melihat jumlah yang ada di dataset

In [13]:
df.isnull().sum() #melihat jika ada value yang null

time(s)        0
V_m            0
Current (A)    0
Cycle          0
Cap            0
dtype: int64

melihat summary keseluruhan di dataset

In [8]:
df.describe() #melihat value dari dataset

time(s)           V_m   Current (A)         Cycle           Cap
count  90982.000000  90982.000000  90982.000000  90982.000000  90982.000000
mean   45490.500000      3.607666      0.999264      4.978699      2.806214
std    26264.385433      0.322777      0.001997      2.582135      0.023283
min        0.000000      2.497000      0.993000      1.000000      2.770180
25%    22745.250000      3.422000      0.997000      3.000000      2.786860
50%    45490.500000      3.636000      0.999000      5.000000      2.804878
75%    68235.750000      3.868000      1.001000      7.000000      2.823520
max    90981.000000      4.114000      1.003000      9.000000      2.840817

merubah nama kolom

In [3]:
df.rename(columns={
    'time(s)': 'time',
    'V_m': 'volt',
    'Current (A)': 'arus',
    'Cycle': 'siklus'
}, inplace=True)
df

time   volt   arus  siklus       Cap
0          0  4.098  0.997       1  2.840817
1          1  4.096  0.998       1  2.840817
2          2  4.094  0.999       1  2.840817
3          3  4.093  0.998       1  2.840817
4          4  4.092  0.999       1  2.840817
...      ...    ...    ...     ...       ...
90977  90977  2.506  1.002       9  2.770180
90978  90978  2.504  1.002       9  2.770180
90979  90979  2.502  1.002       9  2.770180
90980  90980  2.500  1.002       9  2.770180
90981  90981  2.497  1.002       9  2.770180

[90982 rows x 5 columns]

Menghitung hambatan

In [6]:
def R(V, I):
    return(V/I)

Menghitung Daya

In [7]:
def P(V,I):
    return(V*I)

Harus menerapkan konsep looping, conditional if

In [13]:

#def batteray():
#    for siklus in df:
#        if siklus % 2 != 0 :
#            return(volt/arus)
#        elif siklus % 2 == 0:
#            return(volt*arus)
#batteray()


# Pandas Query

melihat voltase dengan kondisi di atas 1.000

In [47]:
voltase = df[df['volt'] > 1.000] #melihat volatase dengan kondisi di atas 1.000
voltase

time   volt   arus  siklus       Cap
0          0  4.098  0.997       1  2.840817
1          1  4.096  0.998       1  2.840817
2          2  4.094  0.999       1  2.840817
3          3  4.093  0.998       1  2.840817
4          4  4.092  0.999       1  2.840817
...      ...    ...    ...     ...       ...
90977  90977  2.506  1.002       9  2.770180
90978  90978  2.504  1.002       9  2.770180
90979  90979  2.502  1.002       9  2.770180
90980  90980  2.500  1.002       9  2.770180
90981  90981  2.497  1.002       9  2.770180

[90982 rows x 5 columns]

melihat volatase dengan kondisi volatase = 4.096 dan kuat arus = 0.998 dan siklus = 7

In [48]:
df[
    (df['volt'] == 4.096) &
    (df['arus'] > 0.998) &
    (df['siklus'] == 7)
]

time   volt  arus  siklus      Cap
60951  60951  4.096   1.0       7  2.78686
60952  60952  4.096   1.0       7  2.78686

# Hasil Akhir

1. Hambatan yang diperoleh dengan sample Volt:2.506 dan Arus:1.002 adalah 5009980039920157

  Daya yang diperoleh dengan sample Volt:2.506 dan Arus:1.002 adalah 5110119999999996

In [8]:
hamabatan =[]
power=[]
for x in df.siklus.unique():
    V=df['volt'][df['siklus'] == x]
    I=df['arus'][df['siklus'] == x]
    if x % 2 != 0:
        hamabatan.append(R(V,I))
    else:
        power.append(P(V,I))

In [9]:
hamabatan

[0        4.110331
 1        4.104208
 2        4.098098
 3        4.101202
 4        4.096096
            ...   
 10232    2.513541
 10233    2.511535
 10234    2.509529
 10235    2.510040
 10236    2.505517
 Length: 10237, dtype: float64,
 20443    4.123370
 20444    4.118236
 20445    4.116232
 20446    4.119358
 20447    4.114228
            ...   
 30613    2.504496
 30614    2.500000
 30615    2.498004
 30616    2.498501
 30617    2.501502
 Length: 10175, dtype: float64,
 40776    4.124373
 40777    4.114114
 40778    4.117234
 40779    4.112112
 40780    4.107000
            ...   
 50878    2.514544
 50879    2.510020
 50880    2.510532
 50881    2.506012
 50882    2.504008
 Length: 10107, dtype: float64,
 60933    4.130653
 60934    4.129648
 60935    4.119358
 60936    4.118355
 60937    4.113226
            ...   
 70969    2.501996
 70970    2.501499
 70971    2.502000
 70972    2.502503
 70973    2.503006
 Length: 10041, dtype: float64,
 81002    4.115115
 81003    4.11000